md-proto/WebAssembly.md

```diff
@@ -0,0 +1,1884 @@
+
+### Function Bodies
+
+Function bodies contain an [array] of [types], which are declarations for
+locals, and a sequence of instructions.
+
+##### Positions within a function body
+
+A *position* within a function refers to an element of the sequence, or to an
+additional *return position* which is sequenced after the end of the function body.
+
+> In the binary encoding, positions are represented as byte offsets; in the text
+format, positions are represented with a special syntax.
+
+
+## Validation
+
+### Module Validation
+
+TODO: describe module validation
+
+### Function Validation
+
+For the duration of the validation of a function body, several data structures
+are created:
+ - A *control-flow stack*, with entries containing
+    - A [control-flow type](#control-flow-type).
+    - A *depth* value.
+ - A *[type] stack*.
+ - A *current position*.
+
+During validation, if either stack is empty when an element is to be popped from
+it, validation fails.
+
+#### Control Flow Type
+
+A *control-flow type* is either a [type](#types), `any`, which satisfies any
+type check, or `void`, which satisfies no check.
+
+> `any` is effectively serving as a kind of bottom type, and `void` is
+effectively serving as a kind of top type. Note that in some type theory
+contexts, it's common to use these names for exactly opposite purposes.
+
+#### Control Flow Type Merge
+
+To merge two [control-flow types]:
+ - If either control-flow type is `any`, the result is the other control flow
+   type.
+ - Otherwise, if the control-flow types are the same, the result is that control
+   flow type.
+ - Otherwise, validation fails with a type error.
+
+#### Function Validation Initialization
+
+The current position starts at the first position. The type stack begins empty.
+The control-flow stack starts with one entry. This entry's control-flow type is
+the first return type of the function, if it has any, or `void` otherwise. Its
+block depth is zero.
+
+#### Function Body Validation
+
+If the current position is the return position, function return validation is
+prompted and validation of the function is thereafter complete.
+
+Otherwise, the instruction at the current position is remembered, the current
+position is incremented to point to the instruction following it, or if there
+are no instructions following it, to the return position. Then the remembered
+instruction is validated as follows.
+
+If the instruction has a **Validation** clause, that clause describes its
+validation. Otherwise, the following generic validation is performed:
+
+For each operand [type] in the instruction's signature, a type is popped from
+the type stack and checked to be the same as the corresponding operand type.
+Each return type of the instruction's signature is then pushed onto the type
+stack.
+
+If the instruction's signature has no return types, and the size of the value
+stack is greater than the control-flow stack's top's depth, validation fails
+with a type error.
+
```

Comment:
This uses the term 'values stack' but before 'type stack'. I don't think this would work at present, as `i32.cont 1, drop, return 0` does not leave an empty type stack at the return point, rather a stack with one `void` type element. CIL seems to have this empty-stack requirement at return points, but that does not make it a good thing to follow, and CIL is a stack machine with other burdens and failed to look ahead to multiple-value support.